src/foam/nanos/geocode/GoogleMapsGeocodingDAO.java

```diff
@@ -0,0 +1,168 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.geocode;
+
+import foam.core.FObject;
+import foam.core.X;
+import foam.dao.DAO;
+import foam.dao.ProxyDAO;
+import foam.lib.json.JSONParser;
+import foam.nanos.auth.Address;
+import foam.nanos.auth.User;
+import foam.util.SafetyUtil;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+public class GoogleMapsGeocodingDAO
+    extends ProxyDAO
+{
+  public static String API_HOST = "https://maps.googleapis.com/maps/api/geocode/json?address=";
+
+  protected String apiKey_;
+  protected JSONParser parser_;
+  protected ThreadLocal<StringBuilder> sb = new ThreadLocal<StringBuilder>() {
+    @Override
+    protected StringBuilder initialValue() {
+      return new StringBuilder();
+    }
+
+    @Override
+    public StringBuilder get() {
+      StringBuilder b = super.get();
+      b.setLength(0);
+      return b;
+    }
+  };
+
+  public GoogleMapsGeocodingDAO(X x, String apiKey, DAO delegate) {
+    setX(x);
+    setDelegate(delegate);
+    this.apiKey_ = apiKey;
+    this.parser_ = getX().create(JSONParser.class);
```

Comment:
Is the JSONParser thread-safe?

Reply:
I don't imagine it is so I will just make this be a local variable.